### PR TITLE
network: TC introduce sfb - Stochastic Fair Blue

### PR DESCRIPTION
**Pull request by ssahani**
Please see https://www.systutorials.com/docs/linux/man/8-tc-sfb/

### CODE REVIEW

File: man/systemd.network.xml [hunk 1]
```diff
@@ -2422,6 +2422,30 @@
...
+        <listitem>
+          <para>Specifies the parent Queueing Discipline (qdisc). Takes one of <literal>root</literal>,
+          <literal>clsact</literal> or <literal>ingress</literal>. Defaults to <literal>root</literal>.</para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>PacketLimit=</varname></term>
+        <listitem>
+          <para>Specifies the hard lmit on the queue size in number of packets. When this limit is reached, incoming packets are
```
Comment (nabijaczleweli) [inline man/systemd.network.xml]: lmit => limit?

Comment (ssahani) [reply man/systemd.network.xml]: ahh thank you . I fixed the other one too.

File: src/network/networkd-network-gperf.gperf [hunk 2]
```diff
@@ -288,6 +288,8 @@ NetworkEmulator.DelayJitterSec,              config_parse_network_emulator_delay
 NetworkEmulator.LossRate,                    config_parse_network_emulator_rate,                       QDISC_KIND_NETEM,              0
 NetworkEmulator.DuplicateRate,               config_parse_network_emulator_rate,                       QDISC_KIND_NETEM,              0
 NetworkEmulator.PacketLimit,                 config_parse_network_emulator_packet_limit,               QDISC_KIND_NETEM,              0
+StochasticFairBlue.Parent,                   config_parse_qdisc_parent,                                QDISC_KIND_SFB,                0
```
Comment (yuwata) [inline src/network/networkd-network-gperf.gperf:291]: Please rebase and add `Handle=`.

File: src/network/tc/sfb.c [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        if (opt.bin_size == 0)
+                opt.bin_size = (opt.max * 4 + 3) / 5;
+
+        r = sd_netlink_message_open_container_union(req, TCA_OPTIONS, "sfb");
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not open container TCA_OPTIONS: %m");
+
+        r = sd_netlink_message_append_data(req, TCA_SFB_PARMS, &opt, sizeof(struct tc_sfb_qopt));
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not append TCA_OPTIONS attribute: %m");
```
Comment (yuwata) [inline src/network/tc/sfb.c]: `TCA_SFB_PARMS`

File: src/network/tc/sfb.c [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        assert(link);
+        assert(qdisc);
+        assert(req);
+
+        sfb = SFB(qdisc);
+
+        if (sfb->packet_limit > 0)
+                opt.limit = sfb->packet_limit;
+
+        if (opt.max == 0 && opt.bin_size >= 1)
```
Comment (yuwata) [inline src/network/tc/sfb.c]: `opt.bin_size` is always zero, currently. Please drop this condition.

File: src/network/tc/sfb.c [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                opt.max = (opt.bin_size * 5 + 1) / 4;
+        else
+                opt.max = 25;
+
+        if (opt.bin_size == 0)
```
Comment (yuwata) [inline src/network/tc/sfb.c]: Please also drop this condition, as this is always true.

File: src/network/tc/sfb.c [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        };
+        int r;
+
+        assert(link);
+        assert(qdisc);
+        assert(req);
+
+        sfb = SFB(qdisc);
+
+        if (sfb->packet_limit > 0)
```
Comment (yuwata) [inline src/network/tc/sfb.c]: This condition can be also dropped.

**Comment by ssahani** [comment]
Updated thanks @yuwata .

**Comment by yuwata** [comment]
Included in #14890. Closing.

**Pull request closed without merging**